### DISCUSSION TITLE
Specify services and topics for rosbridge

### DISCUSSION THREAD

**Discussion by hck007** [RobotWebTools/rosbridge_suite#985] [Q&A] [answered]
I'm using humble version of rosbridge. I'd like to whitelist topics and services for my rosbridge in the ros2 python launch file. There is one XML launch file here https://github.com/RobotWebTools/rosbridge_suite/blob/humble/rosbridge_server/launch/rosbridge_websocket_launch.xml so I inlcude the XML launch file in the ros2 python launch file. I think we need to specify `topics_glob` and `services_glob` to whitelist topics. I tried to pass a list to both `services_glob` and `topics_glob` but rosbrdige just concatenates all elements of the list to a single string value when I verified the parameter `ros2 param get rosbridge_websocket topics_glob`. I then passed a string representation of the list and got the error `[rosbridge_websocket-1] rclpy.exceptions.InvalidParameterTypeException: Trying to set parameter 'services_glob' to '['/service_a', '/service_b']' of type 'STRING_ARRAY', expecting type 'STRING'`.
. Any ideas?
```
import os
import yaml
from ament_index_python.packages import get_package_share_directory
from launch import LaunchDescription
from launch.actions import IncludeLaunchDescription
from launch_xml.launch_description_sources import XMLLaunchDescriptionSource


def generate_launch_description():
    rosbridge_pkg_path = os.path.join(get_package_share_directory("rosbridge_server"),
                                      "launch",
                                      "rosbridge_websocket_launch.xml")
    topics_glob = ["topic_a", "topic_b", "topic_c"]
    services_glob = ["service_a", "service_b", "service_c"]
    rosbridge_launch_arguments = {"port": str(7005),"topics_glob": topics_glob, "services_glob": services_glob}
    rosbridge_include_command = IncludeLaunchDescription(
        XMLLaunchDescriptionSource(rosbridge_pkg_path),
        launch_arguments=rosbridge_launch_arguments.items()
    )
    return LaunchDescription([rosbridge_include_command])
```

**Comment by harsh-pal-robotics** [comment, marked as answer]
You need to pass the paramater as this:

``` python
topics_glob = "'topic_a'",'topic_b','topic_c'"
services_glob = "'service_a','service_b','service_c'"
```

This where the params are being parsed in the rosbridge script:

https://github.com/RobotWebTools/rosbridge_suite/blob/7cd26b582949506ce044c75f0ab22ae02f294868/rosbridge_server/scripts/rosbridge_websocket.py#L282



This helper function can convert an array of strings into the glob param for rosbridge in your launch file:
``` python
def parse_globs(globs: list[str]) -> str:
    return ",".join([f"'{glob}'" for glob in globs])
```

**Reply by hck007** [reply]
Thanks for the answer. The answer is almost correct. After using the method given above, I got the error below. The correct syntax is `topics_glob = "\"'topic_a','topic_b','topic_c'\""`. 
```
[ERROR] [launch]: Caught exception in launch (see debug for traceback): Failed to convert ''topic_a','topic_b','topic_c'' using yaml rules: yaml.safe_load() failed
expected '<document start>', but found ','
  in "<unicode string>", line 1, column 10:
    'topic_a','topic_b','topic_c'
```
